pkg/csi/manila/util/exportlocation.go

```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/gophercloud/gophercloud/openstack/sharedfilesystems/v2/shares"
+)
+
+// Predicate type for filtering out export locations.
+// It's supplied with the index into the export location slice that's being scanned.
+// Should return (true, nil) when search conditions are satisfied,
+// and (false, nil) when they are not.
+type ExportLocationPredicate func(index int) (match bool, err error)
+
+// Predicate matching any export location
+func AnyExportLocation(int) (bool, error) { return true, nil }
+
+// Searches for an export location.
+// Returns index of an export location from the `locs` slice that satisfies following rules:
+// 1. Location is not admin-only and is not empty
+// 2. Location satisfies ExportLocationPredicate
+// Search is biased:
+// 1. Location.Preferred == true is preferred over Location.Preferred == false
+// 2. Locations with lower index are preferred over those with higher index
+func FindExportLocation(locs []shares.ExportLocation, pred ExportLocationPredicate) (index int, err error) {
+	const invalidIdx = -1
+	firstMatchNotPreferred := invalidIdx
```

Comment:
confused, why is the name firstMatchNotPreferred , I guess this stores, first preferred match on contrary

Reply:
A preferred location is `locs[i].Preferred==true`, which makes the function return early:

https://github.com/gman0/cloud-provider-openstack/blob/26613136736920cd667fa28fdc56d8b103cd19fc/pkg/csi/manila/util/exportlocation.go#L51-L55

When not preferred (`Preferred==false`), we keep going through the list, searching for a preferred location. If none is found, we return the first matching "not preferred" location.

Reply:
Ah, got it, you are returning if preferred.